Common/Statistics/TradeBuilder.cs

```diff
@@ -0,0 +1,486 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using QuantConnect.Orders;
+
+namespace QuantConnect.Statistics
+{
+    /// <summary>
+    /// The TradeBuilder class generates trades from executions and market price updates
+    /// </summary>
+    public class TradeBuilder
+    {
+        /// <summary>
+        /// Helper class to manage pending trades and market price updates for a symbol
+        /// </summary>
+        private class Position 
+        {
+            internal List<Trade> PendingTrades { get; set; }
+            internal List<OrderEvent> PendingFills { get; set; }
+            public decimal TotalFees { get; set; }
+            internal decimal MaxPrice { get; set; }
+            internal decimal MinPrice { get; set; }
+
+            public Position()
+            {
+                PendingTrades = new List<Trade>();
+                PendingFills = new List<OrderEvent>();
+            }
+        }
+
+        private readonly List<Trade> _closedTrades = new List<Trade>();
+        private readonly Dictionary<string, Position> _positions = new Dictionary<string, Position>();
```

Comment:
I think we'll need to have this be `Dictionary<Symbol, Position>` so it properly handles mapping changes. Ideally, everywhere in the LEAN code base we'll refer to symbols using the `Symbol` type, only in Tests and the algorithm implementation will we allow the usage of the `string symbol` -- this is because eventually we'll move to a unique symbol id system that may not be symbol base at all (maybe an integer identifier)